boutdata/squashoutput.py

```diff
@@ -215,3 +258,180 @@ def squashoutput(
         # Note that get_chunk_cache() returns a tuple, so we have to unpack it when
         # passing to set_chunk_cache.
         set_chunk_cache(*netcdf4_chunk_cache)
+
+
+class SquashWorkers:
+    """
+    Class for packaging up worker processes for parallel writes, or passing the write
+    through in serial if parallel functionality not requested
+
+    Parameters
+    ----------
+    parallel : bool or int, default False
+        If False, write in serial. If True or 0 use all available processes. If positive
+        integer, use that many processes.
+    filenames : list of str
+        Names of the files to write to.
+    kwargs : dict
+        Keyword arguments to pass to DataFile constructors
+    """
+
+    def __init__(self, parallel, filenames, kwargs):
+        self.parallel = parallel
+        self.kwargs = kwargs
+        if self.parallel is False or len(filenames) == 1:
+            # No point doing parallel writing if there is only one output file
+            self.parallel = False
+            self.open_files(filenames)
+            return
+
+        if self.parallel is True or self.parallel == 0:
+            from boututils.run_wrapper import determineNumberOfCPUs
+
+            self.parallel = determineNumberOfCPUs()
+
+        self.n_outputs = len(filenames)
+
+        self.create_workers(filenames)
+
+    def __del__(self):
+        if self.parallel is False:
+            for f in self.files:
+                f.close()
+                del f
```

Comment:
Would `gc.collect` be needed after this line?

Reply:
I don't think so, but not at all sure. I think it's handled at the moment by having `gc.collect()` after `del workers; del output` at l. 244 above. I have had problems though in particular with the parallelisation classes getting them to clean up all their workers so that they don't hang when a script exits. I don't think putting `gc.collect()` in a `__del__()` method would help, because `__del__()` methods are only called when the garbage collector is invoked (according to this blog post http://foobarnbaz.com/2015/07/09/understanding-python-del-and-garbage-collection/).

Reply:
Ahh...that's a really good point. Thanks for the link :)